src/ir_Argo.cpp

```diff
@@ -0,0 +1,282 @@
+*/
+
+#include "ir_Argo.h"
+#include <algorithm>
+#include "IRremoteESP8266.h"
+#include "IRutils.h"
+#include <Arduino.h>
+
+
+//  ARgo
+
+// Constants
+// using SPACE modulation. MARK is always const 400u
+#define ARGO_PREAMBLE_1           6400U  // Mark
+#define ARGO_PREAMBLE_2           3300U  // Space
+#define ARGO_MARK                  400U
+#define ARGO_ONE_SPACE            2200U
+#define ARGO_ZERO_SPACE            900U
+
+#if SEND_ARGO
+// Send a Argo A/C message.
+//
+// Args:
+//   data: An array of ARGO_COMMAND_LENGTH bytes containing the IR command.
+//
+// Status: STABLE
+//
+// Overloading the IRSend Function
+
+void IRsend::sendArgo(unsigned char data[], uint16_t nbytes,
+                        uint16_t repeat) {
+  if (nbytes < ARGO_COMMAND_LENGTH)
+    return;  // Not enough bytes to send a proper message.
+  // Set IR carrier frequency
+  enableIROut(38);
+  for (uint16_t r = 0; r <= repeat; r++) {
+    // Header
+    // TODO(kaschmo): validate
+    mark(ARGO_PREAMBLE_1);
+    space(ARGO_PREAMBLE_2);
+    // send data, defined in IRSend.cpp
+    for (uint16_t i = 0; i < nbytes; i++)
+      sendData(ARGO_MARK, ARGO_ONE_SPACE, ARGO_MARK,
+               ARGO_ZERO_SPACE, data[i], 8, false);
+               // send LSB first reverses the bit order in array for sending.
+  }
+}
+
+IRArgoESP::IRArgoESP(uint16_t pin) : _irsend(pin) {
+  stateReset();
+}
+
+void IRArgoESP::begin() {
+  _irsend.begin();
+}
+
+void IRArgoESP::send() {
+  Serial.println("Sending IR code");
+  _irsend.sendArgo(argo);
+}
+
+void IRArgoESP::checksum() {
+  uint8_t sum = 2; // Corresponds to byte 11 being constant 0b01
+  uint8_t i;
+
+  // Only add up bytes to 9. byte 10 is 0b01 constant anyway.
+  // Assume that argo array is MSB first (left)
+  for (i = 0; i < 10; i++)
+    sum += argo[i];
+
+  sum = sum % 256;  // modulo 256
+  // Append sum to end of array
+  // Set const part of checksum bit 10
+  argo[10] = 0b00000010;
+  argo[10] += sum << 2;  // Shift up 2 bits and append to byte 10
+  argo[11] = sum >> 6;  // Shift down 6 bits and add in two LSBs of bit 11
+}
+
+void IRArgoESP::stateReset() {
+  for (uint8_t i = 0; i < ARGO_COMMAND_LENGTH; i++)
+    argo[i] = 0x0;
+
+  // Argo Message. Store MSB left.
+  // Default message:
+  argo[0] = 0b10101100;  // LSB first (as sent) 0b00110101; //const preamble
+  argo[1] = 0b11110101;  // LSB first: 0b10101111; //const preamble
+  // Keep payload 2-9 at zero
+  argo[10] = 0b00000010;  // Const 01, checksum 6bit
+  argo[11] = 0b00000000;  // Checksum 2bit
+
+  this->off();
+  this->setTemp(20);
+  this->setRoomTemp(25);
+  this->setCoolMode(ARGO_COOL_AUTO);
+  this->setFan(ARGO_FAN_AUTO);
+
+  checksum();
+}
+
+uint8_t* IRArgoESP::getRaw() {
+  checksum();  // Ensure correct settings before sending.
+  return argo;
+}
+
+void IRArgoESP::on() {
+  // state = ON;
+  ac_state=1;
+  // Bit 5 of byte 9 is on/off
+  // in MSB first
+  argo[9] = argo[9] | 0b00100000;  // Set ON/OFF bit to 1
+
+  checksum();
```

Comment:
Took this pattern over from other A/C implementations. But since I am not storing the actual state in the bit array  you are right. Checksum is only required just before sending.